scripts/download_data.py

```diff
@@ -0,0 +1,290 @@
+"""
+Constructs a data source for the ga4gh server by downloading data from
+authoritative remote servers.
+"""
+# TODO
+# - need some kind of checkpoint functionality to resume process
+#   where it left off since getting a clean run is so rare...
+# - references support
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import argparse
+import glob
+import gzip
+import os
+import tempfile
+import urllib2
+
+import pysam
+
+import utils
+
+
+def mkdirAndChdirList(dirNameList):
+    """
+    For each entry in dirNameList, make a directory if it does not exist and
+    cd into it
+    """
+    for directory in dirNameList:
+        mkdirAndChdir(directory)
+
+
+def mkdirAndChdir(dirName):
+    """
+    Make a directory if it does not exist and cd into it
+    """
+    if not os.path.exists(dirName):
+        os.mkdir(dirName)
+    os.chdir(dirName)
+
+
+def cleanDir():
+    """
+    Removes genomic files in current directory
+    """
+    cwd = os.getcwd()
+    utils.log("Cleaning out directory '{}'".format(cwd))
+    globs = [
+        "*.tbi", "*.vcf", "*.vcf.gz", "*.bam", "*.bam.bai", "*.fasta.gz",
+        "*.fasta", "*.fasta.fai"]
+    for fileGlob in globs:
+        fileNames = glob.glob(fileGlob)
+        for fileName in fileNames:
+            os.remove(fileName)
+
+
+def escapeDir(levels=4):
+    # back to orig dir
+    for _ in range(levels):
+        os.chdir('..')
+
+
+class ChromMinMax(object):
+    """
+    A container class for storing the min and max position seen
+    for every chromosome
+    """
+    defaultMinPos = 2**30
+    defaultMaxPos = 0
+
+    class MinMax(object):
+        def __init__(self):
+            self.minPos = ChromMinMax.defaultMinPos
+            self.maxPos = ChromMinMax.defaultMaxPos
+
+    def __init__(self):
+        self.chromMap = {}
+
+    def addPos(self, chrom, position):
+        if chrom not in self.chromMap:
+            self.chromMap[chrom] = self.MinMax()
+        minMax = self.chromMap[chrom]
+        if position > minMax.maxPos:
+            minMax.maxPos = position
+        if position < minMax.minPos:
+            minMax.minPos = position
+
+    def getMinPos(self, chrom):
+        minMax = self.chromMap[chrom]
+        return minMax.minPos
+
+    def getMaxPos(self, chrom):
+        minMax = self.chromMap[chrom]
+        return minMax.maxPos
+
+
+class AbstractFileDownloader(object):
+    """
+    Base class for individual site genome file downloaders
+    """
+    def __init__(self, args):
+        self.args = args
+        self.datasetId = 'dataset1'
```

Comment:
I guess technically this should be ``datasetName`` now, as the IDs are server defined. Likewise for the variantSet and ReferenceSet.